django_project/certification/views/certificate.py

```diff
@@ -489,3 +490,78 @@ def top_up_unavailable(request, **kwargs):
         context={
             'the_project': project,
             'has_pending_organisations': has_pending})
+
+
+def email_all_attendees(request, **kwargs):
+    project_slug = kwargs.get('project_slug', None)
+    course_slug = kwargs.get('course_slug', None)
+    organisation_slug = kwargs.get('organisation_slug', None)
+    project = Project.objects.get(slug=project_slug)
+    course = Course.objects.get(slug=course_slug)
+    attendee_list = \
+        Certificate.objects.filter(
+            is_paid=True, course=course).values_list('attendee', flat=True)
+    attendee_list_object = []
+    for attendee_pk in attendee_list:
+        attendee = Attendee.objects.get(pk=attendee_pk)
+        attendee_list_object.append(attendee)
+
+    organisation = CertifyingOrganisation.objects.filter(approved=False)
+    has_pending = False
+    if organisation:
+        has_pending = True
+
+    url = reverse('course-detail', kwargs={
+        'project_slug': project_slug,
+        'organisation_slug': organisation_slug,
+        'slug': course_slug
+    })
+
+    if request.method == 'POST':
+
+        site = request.get_host()
+        for attendee in attendee_list_object:
+            # Send email to each attendee with the link to his certificate.
+
+            send_mail(
+                'Certificate from %s Course' % course.course_type,
+                'Dear %s %s,\n\n' % (
```

Comment:
Your code will be much cleaner and more intuitive if you use a dictionary for the string replacements.

See http://thomas-cokelaer.info/blog/2011/05/python-string-substitution-using-dictionaries/ for an example.

Will you make a follow up PR that does that and adopt it as standard practice for any string with more than two or three substitions in future?

Reply:
Will do

Reply:
Your link Tim is still the old way for string formatting in Python.
Please use the `format` function: https://pyformat.info `Named placeholders` section.

Even for one substitution, we use a dictionary in InaSAFE, because for translators in Transifex, it's far away better to always have a placeholder.